src/app/reporting/Engine.cpp

```diff
@@ -223,6 +223,74 @@ bool IsClusterDataVersionEqualTo(DataModel::Provider * dataModel, const Concrete
     return info.has_value() && (info->dataVersion == dataVersion);
 }
 
+/// Check if the given `err` is a known ACL error that can be translated into
+/// a StatusIB (UnsupportedAccess/AccessRestricted)
+///
+/// Returns true if the error could be translated and places the result into `outStatus`.
+/// `path` is used for logging.
+bool IsTranslatableAclError(const ConcreteEventPath & path, const CHIP_ERROR & err, StatusIB & outStatus)
+{
+    if ((err != CHIP_ERROR_ACCESS_DENIED) && (err != CHIP_ERROR_ACCESS_RESTRICTED_BY_ARL))
+    {
+        return false;
+    }
+
+    ChipLogDetail(InteractionModel, "Access to event (%u, " ChipLogFormatMEI ", " ChipLogFormatMEI ") denied by %s",
+                  path.mEndpointId, ChipLogValueMEI(path.mClusterId), ChipLogValueMEI(path.mEventId),
+                  err == CHIP_ERROR_ACCESS_DENIED ? "ACL" : "ARL");
+
+    outStatus = err == CHIP_ERROR_ACCESS_DENIED ? StatusIB(Status::UnsupportedAccess) : StatusIB(Status::AccessRestricted);
+    return true;
+}
+
+CHIP_ERROR CheckEventValidity(const ConcreteEventPath & path, const SubjectDescriptor & subjectDescriptor,
+                              DataModel::Provider * provider, StatusIB & outStatus)
+{
+    // We validate ACL before Path, however this means we do not want the real ACL check
+    // to be blocked by a `Invalid endpoint id` error when checking event info.
+    // As a result, we check for VIEW privilege on the cluster first (most permissive)
+    // and will do a 2nd check for the actual required privilege as a followup.
+    RequestPath requestPath{
+        .cluster     = path.mClusterId,
+        .endpoint    = path.mEndpointId,
+        .requestType = RequestType::kEventReadRequest,
+    };
+    CHIP_ERROR err = GetAccessControl().Check(subjectDescriptor, requestPath, Access::Privilege::kView);
+    if (IsTranslatableAclError(path, err, outStatus))
+    {
+        return CHIP_NO_ERROR;
+    }
+    ReturnErrorOnFailure(err);
+
+    DataModel::EventEntry eventInfo;
+    err = provider->EventInfo(path, eventInfo);
+    if (err != CHIP_NO_ERROR)
+    {
+        outStatus = StatusIB(err);
```

Comment:
This neither matches the old code nor the spec, right?  In the sense that what the spec has here is returning the correct one of UNSUPPORTED_ENDPOINT/CLUSTER/EVENT, and we were doing the right thing for the endpoint/cluster and had no concept of UNSUPPORTED_EVENT.

Is `provider->EventInfo()` returning the relevant IM status codes?  Its certainly not documented to, or documented at all.

As in, the ValidateClusterPath below is pointless, because if cluster instance does not exist we will lose here already, no?

I guess this is "working" right now because the codegen provider always succeeds on `EventInfo`?  But again, the API does not say that's the behavior....